### PR TITLE
fix(auth/oauth2): Fix code_verifier generation in oauth2 flow.

### PR DESCRIPTION
**Pull request by jtk54**

### CODE REVIEW

**Comment by jtk54** [comment]
Don't merge, there's an issue with refreshing the token with a code_verifier.

**Comment by jtk54** [comment]
Nevermind, this is fine. False signal from changes I had in Gate.

**Comment by jtk54** [comment]
I have debug changes in this branch as well, I'll ping when this is cleaned up.

File: cmd/gateclient/client.go [hunk 1]
```diff
@@ -294,20 +295,19 @@ func (m *GatewayClient) Authenticate() error {
...
-			if _, err := rand.Read(verifierBytes); err != nil {
-				util.UI.Error("Could not generate random string for code_verifier")
+			verifier, err := generateCodeVerifier()
+			if err != nil {
 				return err
 			}
-			verifier := string(verifierBytes)
+
 			codeChallenge := oauth2.SetAuthURLParam("code_challenge", verifier)
 			codeVerifier := oauth2.SetAuthURLParam("code_verifier", verifier)
```
Comment (ezimanyi) [inline cmd/gateclient/client.go]: Should we be doing a sha256 hash of this given that we're capable?  Per the RFC "If the client is capable of using "S256", it MUST use "S256", as "S256" is Mandatory To Implement (MTI) on the server."

File: cmd/gateclient/client.go [hunk 2]
```diff
@@ -323,6 +365,25 @@ func (m *GatewayClient) Authenticate() error {
 	return nil
 }
 
+// generateCodeVerifier generates an OAuth2 code verifier
+// in accordance to https://www.oauth.com/oauth2-servers/pkce/authorization-request/.
+func generateCodeVerifier() (string, error) {
+	randomBytes := make([]byte, 64)
+	if _, err := rand.Read(randomBytes); err != nil {
+		util.UI.Error("Could not generate random string for code_verifier")
+		return "", err
```
[the remark this answers is not in the document]
Comment (jtk54) [reply cmd/gateclient/client.go]: Strings can't be nil.

File: cmd/gateclient/client.go (hunk 1 above)
Comment (jtk54) [reply cmd/gateclient/client.go]: Makes sense.

File: cmd/gateclient/client.go [hunk 3]
```diff
@@ -323,6 +365,25 @@ func (m *GatewayClient) Authenticate() error {
...
+	}
+
+	const letters = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-._~"
+	verifierBytes := make([]byte, 64)
+	for i, b := range randomBytes {
+		verifierBytes[i] = letters[b%byte(len(letters))]
```
[the remark this answers is not in the document]
Comment (jtk54) [reply cmd/gateclient/client.go]: Makes sense.

File: cmd/gateclient/client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ezimanyi) [reply cmd/gateclient/client.go]: :+1:...that's a pretty good reason

**Pull request merged**